### PR TITLE
Fix JSON deserialization of AssignExpr

### PR DESCRIPTION
**Pull request by overlookmotel**
**Description:**

Fix for #5168. However:

1. Surely there's an easier way to do this?
2. I imagine there is a more idiomatic way to use `match { }` here:

https://github.com/overlookmotel/swc/blob/assign-expr-deserialize/crates/swc_ecma_ast/src/expr.rs#L568-L576

I would appreciate any pointers as to how to improve this.

**Related issue (if exists):**

#5168

### CODE REVIEW

**Review by kdy1 — CHANGES_REQUESTED** [review]
Please remove images

**Comment by magic-akari** [comment]
Is there a case where `PatOrExpr::Expr` is not `Expr::Ident`? @kdy1 
I wonder if `PatOrExpr` is redundant?

**Comment by magic-akari** [comment]
It seems a lot of work to do if I remove the `PatOrExpr`, I will try it later.
And I don't have strong opinions on this pull request change. It's ok to fix #5168.

File: crates/swc_ecma_ast/src/expr.rs [hunk 1]
```diff
@@ -471,6 +492,102 @@ impl Take for AssignExpr {
...
+                if op_field.is_none() {
+                    return Err(de::Error::missing_field("operator"));
+                }
+                if left_field.is_none() {
+                    return Err(de::Error::missing_field("left"));
+                }
+                if right_field.is_none() {
+                    return Err(de::Error::missing_field("right"));
+                }
+
```
Comment (kdy1) [inline crates/swc_ecma_ast/src/expr.rs:562]: Can you use `.ok_or_else` instead for these 4 vars?

File: crates/swc_ecma_ast/src/expr.rs [hunk 2]
```diff
@@ -471,6 +492,102 @@ impl Take for AssignExpr {
...
+            {
+                let mut span_field: Option<Span> = None;
+                let mut op_field: Option<AssignOp> = None;
+                let mut left_field: Option<PatOrExpr> = None;
+                let mut right_field: Option<Box<Expr>> = None;
+
+                while let Some(key) = map.next_key::<&str>()? {
+                    match key {
+                        "span" => {
+                            if Option::is_some(&span_field) {
```
Comment (kdy1) [inline crates/swc_ecma_ast/src/expr.rs:521]: Also, please change this to `span_field.is_some()`

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you!

File: crates/swc_ecma_ast/src/expr.rs (hunk 1 above)
Comment (overlookmotel) [reply crates/swc_ecma_ast/src/expr.rs]: Done.

File: crates/swc_ecma_ast/src/expr.rs (hunk 2 above)
Comment (overlookmotel) [reply crates/swc_ecma_ast/src/expr.rs]: Done.

File: crates/swc_ecma_ast/src/expr.rs [hunk 3]
```diff
@@ -471,6 +492,102 @@ impl Take for AssignExpr {
...
+                                return Err(de::Error::duplicate_field("span"));
+                            }
+                            span_field = Some(map.next_value()?);
+                        }
+                        "operator" => {
+                            if Option::is_some(&op_field) {
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/swc_ecma_ast/src/expr.rs]: Done.

File: crates/swc_ecma_ast/src/expr.rs [hunk 4]
```diff
@@ -471,6 +492,102 @@ impl Take for AssignExpr {
...
+                                return Err(de::Error::duplicate_field("operator"));
+                            }
+                            op_field = Some(map.next_value()?);
+                        }
+                        "left" => {
+                            if Option::is_some(&left_field) {
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/swc_ecma_ast/src/expr.rs]: Done.

File: crates/swc_ecma_ast/src/expr.rs [hunk 5]
```diff
@@ -471,6 +492,102 @@ impl Take for AssignExpr {
...
+                                return Err(de::Error::duplicate_field("left"));
+                            }
+                            left_field = Some(map.next_value()?);
+                        }
+                        "right" => {
+                            if Option::is_some(&right_field) {
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/swc_ecma_ast/src/expr.rs]: Done.

**Review by kdy1 — APPROVED** [review]
Thanks!

---

swc-bump:
 - swc_ecma_ast

**Comment by overlookmotel** [comment]
I've made the changes requests and tidied up some other bits.

I'm still not completely happy with this:

```rust
if let PatOrExpr::Pat(ref pat) = left {
  if let Pat::Ident(ident) = &**pat {
    left = PatOrExpr::Expr(Box::new(Expr::Ident(ident.id.clone())));
  }
}
```

The `.clone()` call should be unnecessary - `ident.id` can just be moved from the old structure to the new. But I couldn't find a way to get the compiler to accept this, since `pat` is borrowed.

My first attempt was:

```rust
left = match left {
  PatOrExpr::Pat(pat) => match *pat {
    Pat::Ident(ident) => PatOrExpr::Expr(Box::new(Expr::Ident(ident.id))),
    pat => PatOrExpr::Pat(Box::new(pat)),
  },
  left => left,
}
```

This avoids cloning as it takes ownership, but has the worse consequences of unclear code and re-creating `PatOrExpr::Pat(Box::new(pat))` unnecessarily in the majority of cases.

Or am I worrying too much? Will the compiler compile away the `.clone()` call?

**Comment by kdy1** [comment]
I think so?
Even if the compiler fails to optimize it, it will not affect performance in almost all cases.
`clone()` of `Ident` is typically very cheap.

**Comment by overlookmotel** [comment]
Is there no way to tell compiler "give me ownership of `ident`, and make `left` and `pat` uninitialized"? Sorry, this has nothing to do with SWC specifically, I'm just still learning Rust. This feels like it'd be a common situation.

**Pull request merged**